### PR TITLE
UHMAEAT 4.0 (main/period)times

### PR DESCRIPTION
**Pull request by wonderingabout**
major optimizations and simplifications over the old code

edit : you can also see the changes here directly : 
https://github.com/wonderingabout/gtp2ogs/blob/UHMAEAT-4-0/connection.js#L651

edit 2 : now tested and ready to merge :+1:

### CODE REVIEW

File: connection.js [hunk 1]
```diff
@@ -855,158 +830,131 @@ class Connection {
...
                 timeNotificationToString = timespanToDisplayString(t.total_time);
-                endingSentence = ".";
                 return UHMAEATForMaintimesGenericReject();
             }
+
+            // FINAL UNIVERSAL REJECT :
+            function UHMAEATForMaintimesGenericReject() {
+                conn_log(`${user.username} wanted ${argIsMinMax.minimumMaximum} ${timecontrolSentence} ${t.speed} ${rankedUnranked} ${timeNotificationToString} in ${t.time_control}, it is below ${timeToString}`);
+                // example : "userX wanted minimum maintime blitz ranked 5 seconds,
+                // it is below 1 minutes"
```
Comment (Dorus) [inline connection.js]: The example doesn't seem to match the code, ` in ${t.time_control}` is nowhere to be found in the examlpe.

File: connection.js [hunk 2]
```diff
@@ -855,158 +830,131 @@ class Connection {
...
+                return { reject : true, msg: `${argIsMinMax.minimumMaximum} ${timecontrolSentence} for ${t.speed} ${rankedUnranked} ${timeToString}, please ${argIsMinMax.increaseReduce} ${timecontrolSentence}.` };
+                // example : "minimum maintime for blitz ranked games is 5 seconds,
```
Comment (Dorus) [inline connection.js]: Is ` is ` in the actual code?

Comment (wonderingabout) [reply connection.js]: indeed ! 
fixed :+1:

File: connection.js (hunk 1 above)
Comment (wonderingabout) [reply connection.js]: indeed ! 
fixed :+1:

File: connection.js [hunk 3]
```diff
@@ -676,175 +678,148 @@ class Connection {
...
             }
+
+            // FINAL UNIVERSAL REJECT :
+            function UHMAEATForPeriodtimesGenericReject() {
+                conn_log(`${user.username} wanted ${argIsMinMax.minimumMaximum} ${timecontrolSentence} ${t.speed} ${rankedUnranked} ${timeNotificationToString} in ${t.time_control}, it is below ${timeToString}`);
+                // example : "userX wanted minimum periodtime blitz ranked 5 seconds,
+                // it is below 1 minutes"
+                return { reject : true, msg: `${argIsMinMax.minimumMaximum} ${timecontrolSentence} for ${t.speed} ${rankedUnranked} ${timeToString}, please ${argIsMinMax.increaseReduce} ${timecontrolSentence}.` };
+                // example : "minimum periodtime for blitz ranked games is 5 seconds, 
+                // please increase periodtime"
```
[the remark this answers is not in the document]
Comment (wonderingabout) [reply connection.js]: indeed ! 
fixed :+1:

File: connection.js [hunk 4]
```diff
@@ -676,175 +678,148 @@ class Connection {
...
+                timecontrolSentence = "Time per move";
                 timeNotificationToString = timespanToDisplayString(t.period_time);
-                endingSentence = ".";
                 return UHMAEATForPeriodtimesGenericReject();
             }
+
+            // FINAL UNIVERSAL REJECT :
+            function UHMAEATForPeriodtimesGenericReject() {
+                conn_log(`${user.username} wanted ${argIsMinMax.minimumMaximum} ${timecontrolSentence} ${t.speed} ${rankedUnranked} ${timeNotificationToString} in ${t.time_control}, it is below ${timeToString}`);
+                // example : "userX wanted minimum periodtime blitz ranked 5 seconds,
```
[the remark this answers is not in the document]
Comment (wonderingabout) [reply connection.js]: indeed ! 
fixed :+1:

File: connection.js (hunk 2 above)
Comment (wonderingabout) [reply connection.js]: my messages were a big mess
i rewrote them yesterday and didnt properly verify
thanks for your precise checking :)
should be all good now

**Comment by wonderingabout** [comment]
new version:

```
            function UHMAEATTimecontrolsConditionProcessing(minArgNumber, maxArgNumber) {
                const timecontrolsConditions = 
                      [["fischer", "Initial Time and/or Max Time", t.initial_time, (t.initial_time < uhmaeatTimes.argNumber) || (t.max_time < uhmaeatTimes.argNumber), (t.initial_time > uhmaeatTimes.argNumber) || (t.max_time > uhmaeatTimes.argNumber)],
                      ["byoyomi", "Main Time", t.main_time, t.main_time < uhmaeatTimes.argNumber, t.main_time > uhmaeatTimes.argNumber],
                      ["canadian", "Main Time", t.main_time, t.main_time < uhmaeatTimes.argNumber, t.main_time > uhmaeatTimes.argNumber],
                      ["absolute", "Total Time", t.total_time, t.total_time < uhmaeatTimes.argNumber, t.total_time > uhmaeatTimes.argNumber]];

                let timeCondition = false;
                for (let e of timecontrolsConditions) {
                if (maxArgNumber > -1) {
                    timeCondition = e[4];
                } else if (minArgNumber > -1) {
                    timeCondition = e[3];
                } if (t.time_control === e[0] && timeCondition) {
                    uhmaeatTimes.notificationToString = timespanToDisplayString(e[2]);
                    conn_log(`${user.username} wanted ${argIsMinMax.minimumMaximum} ${e[1]} ${t.speed} ${rankedUnranked} in ${t.time_control} ${uhmaeatTimes.notificationToString}, it is ${argIsMinMax.belowAbove} ${uhmaeatTimes.argToString}`);
                    // example : "userX wanted minimum Main Time blitz ranked
                    // in byoyomi 15 seconds, it is below 1 minutes"
                    return { reject : true, msg: `${argIsMinMax.minimumMaximum} ${e[1]} for ${t.speed} ${rankedUnranked} games in ${t.time_control} is ${uhmaeatTimes.argToString}, please ${argIsMinMax.increaseReduce} ${e[1]}.` };
                    // example : "minimum Main Time for blitz ranked games
                    // in byoyomi is 1 minutes, please increase Main Time."
                }
            }
```

aim is to merge min and max, as well as having a smarter function

**Comment by wonderingabout** [comment]
there was a small mistake btw in periodtimes in the uhmaeat 3.1 and below, period time in fischer is `"increment time"` and `t.increment_time`, not "initial and/or max time"`, fixed in 4.0

File: connection.js [hunk 5]
```diff
@@ -651,411 +630,140 @@ class Connection {
...
-                    return UHMAEATForMaintimesAbsoluteRejectProcessing();
+                    const timecontrolsSettings = timecontrolsMainPeriodTime(mainPeriodTime, argNumber);
+
+                    let timeCondition = false;
+                    for (let setting of timecontrolsSettings) {
+                        if (familyArray[6]) {            // min
+                            timeCondition = setting[3];  // min condition
+                        } else if (familyArray[7]) {     // max
+                            timeCondition = setting[4];  // max condition
+                        } if (t.time_control === setting[0] && timeCondition) {
```
Comment (roy7) [inline connection.js]: This if should be on the next line by itself. It isn't connected to the prior if statement.

**Comment by wonderingabout** [comment]
yeah !
just rebased cleanly !

**Comment by wonderingabout** [comment]
i'm amazed at how nicely it worked at first try 

TODO : change check all Three message in config.js (if all AND ranked AND unranked are used (because defaults)), we will ignore the general value and use specific value if inputted by bot admin

testg11
testg10
testg8
testg7
testg6
testg5
testg4
testg3
testg2
testg1

**Comment by wonderingabout** [comment]
ready to merge now :)

**Pull request merged**